src/main/scala/org/ergoplatform/ErgoAddress.scala

```diff
@@ -0,0 +1,217 @@
+  * - A prefix of address is showing network and an address type.
+  * - An address is using an encoding (namely, Base58) which is avoiding similarly l0Oking characters, friendly to
+  * double-clicking and line-breaking in emails.
+  *
+  *
+  *
+  * An address is encoding network type, address type, checksum, and enough information to watch for a particular scripts.
+  *
+  * Possible network types are:
+  * Mainnet - 0x00
+  * Testnet - 0x10
+  *
+  * Address types are, semantics is described below:
+  * 0x01 - Pay-to-PublicKey(P2PK) address
+  * 0x02 - Pay-to-Script-Hash(P2SH)
+  * 0x03 - Pay-to-Script(P2S)
+  *
+  * For an address type, we form content bytes as follows:
+  *
+  * P2PK - serialized (compressed) public key
+  * P2sH - 192 bit of the script
+  * P2S  - serialized script
+  *
+  * Address examples for testnet:
+  *
+  * 3   - P2PK (3WvsT2Gm4EpsM9Pg18PdY6XyhNNMqXDsvJTbbf6ihLvAmSb7u5RN)
+  * 8   - P2SH (8UmyuJuQ3FS9ts7j72fn3fKChXSGzbL9WC, 8LnSX95GAWdbDZWJZQ73Uth4uE8HqN3emJ)
+  * ?   - P2S (imdaM2NzX, z4hAmfvfSnQJPChMWzfBzJjpB8ei2HoLCZ2RHTaNArMNHFirdJTc7E)
+  *
+  * for mainnet:
+  *
+  * 9  - P2PK (9fRAWhdxEsTcdb8PhGNrZfwqa65zfkuYHAMmkQLcic1gdLSV5vA)
+  * 2  - P2SH (25qGdVWg2yyYho8uC1pLtc7KxFn4nEEAwD, 23NL9a8ngN28ovtLiKLgHexcdTKBbUMLhH)
+  * ?  - P2S (7bwdkU5V8, BxKBaHkvrTvLZrDcZjcsxsF7aSsrN73ijeFZXtbj4CXZHHcvBtqSxQ)
+  *
+  *
+  * Prefix byte = network type + address type
+  *
+  * checksum = blake2b256(prefix byte ++ content bytes)
+  *
+  * address = prefix byte ++ content bytes ++ checksum
+  *
+  */
+
+sealed trait ErgoAddress {
+  val addressTypePrefix: Byte
+
+  val contentBytes: Array[Byte]
+
+  val script: Value[SBoolean.type]
+}
+
+class P2PKAddress(val pubkey: ProveDlog,
+                  val pubkeyBytes: Array[Byte])
+                 (implicit val encoder: ErgoAddressEncoder) extends ErgoAddress {
+
+  override val addressTypePrefix: Byte = P2PKAddress.addressTypePrefix
+
+  override val contentBytes: Array[Byte] = pubkeyBytes
+
+  override val script: ProveDlog = pubkey
+
+  override def equals(obj: Any): Boolean = obj match {
+    case p2pk: P2PKAddress => util.Arrays.equals(pubkeyBytes, p2pk.pubkeyBytes)
+    case _ => false
+  }
+
+  override def hashCode(): Int = Ints.fromByteArray(pubkeyBytes.takeRight(4))
+
+  override def toString: String = encoder.toString(this)
+}
+
+object P2PKAddress {
+  val addressTypePrefix: Byte = 1: Byte
+
+  def apply(pubkey: ProveDlog)(implicit encoder: ErgoAddressEncoder): P2PKAddress = {
+    val bs = ValueSerializer.serialize(pubkey)
+    new P2PKAddress(pubkey, bs)
+  }
+}
+
+class Pay2SHAddress(val scriptHash: Array[Byte])(implicit val encoder: ErgoAddressEncoder) extends ErgoAddress {
+  override val addressTypePrefix: Byte = Pay2SHAddress.addressTypePrefix
+
+  override val contentBytes: Array[Byte] = scriptHash
+
+  //similar script checked in "P2SH - 160 bits" test in sigma repository, but here we use 192 bits
```

Comment:
@greenhat let's clarify this comment. I don't understand this.

Reply:
The comment references http://github.com/ScorexFoundation/sigmastate-interpreter/blob/i230-add-pk/src/test/scala/sigmastate/utxo/ErgoLikeInterpreterSpecification.scala#L358
test, which I suppose was written before the address implementation. I've added the test name to the comment.